apm-agent-core/src/main/java/co/elastic/apm/agent/bci/IndyBootstrap.java

```diff
@@ -370,12 +376,14 @@ public static ConstantCallSite bootstrap(MethodHandles.Lookup lookup,
     }
 
     private static List<String> getClassNamesFromBundledPlugin(String adviceClassName, ClassLoader adviceClassLoader) throws IOException, URISyntaxException {
-        List<String> pluginClasses;
-        String packageName = adviceClassName.substring(0, adviceClassName.lastIndexOf('.'));
-        pluginClasses = classesByPackage.get(packageName);
+        if (!adviceClassName.startsWith(EMBEDDED_PLUGINS_PACKAGE_PREFIX)) {
+            throw new IllegalArgumentException("invalid advice class location : " + adviceClassName);
+        }
+        String pluginPackage = adviceClassName.substring(0, adviceClassName.indexOf('.', EMBEDDED_PLUGINS_PACKAGE_PREFIX.length()));
+        List<String> pluginClasses = classesByPackage.get(pluginPackage);
         if (pluginClasses == null) {
-            classesByPackage.putIfAbsent(packageName, PackageScanner.getClassNames(packageName, adviceClassLoader));
-            pluginClasses = classesByPackage.get(packageName);
+            classesByPackage.putIfAbsent(pluginPackage, PackageScanner.getClassNames(pluginPackage, adviceClassLoader));
```

Comment:
Wouldn't this scan ALL agent classes?

Reply:
I don't think so:

```java
String adviceClassName = "co.elastic.apm.agent.plugin.foo.PluginInstrumentation";
adviceClassName.substring(0, adviceClassName.indexOf('.', "co.elastic.apm.agent.".length()));
```

Returns `co.elastic.apm.agent.plugin`, so it would scan all classes in this package and sub-package.

Reply:
Ahh, of course! Read too quickly, missed the "from-index" search